Storage: STP for new scratch space SC selection logic (#86)
RedHatQE/openshift-virtualization-tests-design-docs: stps/sig-storage/scratch_space_sc_selection_logic.md

# Openshift-virtualization-tests Test plan

## **Scratch Space Storage Class Selection Logic - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/containerized-data-importer/pull/4054
- **Feature Tracking:** https://issues.redhat.com/browse/CNV-72238
- **Epic Tracking:** https://issues.redhat.com/browse/CNV-79031

> - **Feature Maturity:** GA: 4.22

Author: done

    - Existing administrator configurations for scratch space storage class must continue to work unchanged

- [x] **Understand Value and Customer Use Cases**
  - *Describe the feature's value to customers:*
    - Ensures consistent storage provisioning behavior, reducing confusion and configuration errors
    - Aligns temporary resource allocation with production storage requirements
  - *List the customer use cases identified:*
    - As a VM owner, I want scratch space to use the same storage class as my disk so that provisioning is consistent

- [x] **Testability**
  - *Note any requirements that are unclear or untestable:* None - all requirements are testable through CDI operations (clone, import, upload) with various storage class configurations

- [x] **Acceptance Criteria**
  - *List the acceptance criteria:*
    - When no explicit scratch space storage class is configured, scratch space uses the same storage class as the target PVC
    - When an explicit scratch space storage class is configured, scratch space uses that configured storage class

>     - Automated tests verify both default (same-as-target) and explicitly configured scratch space storage class behaviors

Author: changed

  - *Details:*
    - All functional tests will be automated

- [x] **Regression Testing** — Verifies that new changes do not break existing functionality
  - *Details:*
    - Verify existing CDI operations work with new scratch space logic
    - Confirm scratchSpaceStorageClass config continues to work as before (override behavior unchanged)

**Non-Functional**

- [x] **Performance Testing** — Validates feature performance meets requirements (latency, throughput, resource usage)
  - *Details:* N/A - no impact on provisioning performance

- [x] **Scale Testing** — Validates feature behavior under increased load and at production-like scale (e.g., large number of VMs, nodes, or concurrent operations)
  - *Details:* N/A - Selection logic scales with existing CDI controller capabilities

- [x] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization, and vulnerability scanning
  - *Details:* N/A - no new security requirements

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - *Details:* N/A - Backend storage class selection logic only

> 

Author: done

# Openshift-virtualization-tests Test plan

## **Scratch Space Storage Class Selection Logic - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/containerized-data-importer/pull/4054
- **Feature Tracking:** https://issues.redhat.com/browse/CNV-72238
- **Epic Tracking:** https://issues.redhat.com/browse/CNV-79031
- **Feature Maturity:**
  - DP: N/A
  - TP: N/A
  - GA: 4.22
- **QE Owner(s):** Kate Shvaika (kshvaika@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage

**Document Conventions (if applicable):** N/A

### **Feature Overview**

> Some VM disk provisioning operations require temporary scratch space. This feature changes the default storage class selection for scratch space: it now uses the same storage class as the target disk, instead of falling back to the cluster default. Administrators can configure a specific storage class for scratch space to override this behavior.

Reviewer: ```suggestion
Some volume provisioning operations (such as registry imports, image uploads, or certain http imports) require an additional temporary PVC known as a scratch space PVC to act as an intermediary step in processing the data prior to writing it to the target PVC. This feature changes the default storage class selection for scratch space PVCs: it now uses the same storage class as the target volume, instead of falling back to the cluster default. Administrators can still configure a specific scratch space storage class to override this behavior by using the CDI config field `scratchSpaceStorageClass` 
```

Author: Done.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**
  - *List the key D/S requirements reviewed:*

>     - When no scratch space storage class is configured, scratch space must use the same storage class as the target disk

Reviewer: ```suggestion
    - When no scratch space storage class is configured, scratch space must use the same storage class as the target volume
```

Author: Done.

    - When a scratch space storage class is explicitly configured by the administrator, it must override the default same-as-target behavior
    - Existing administrator configurations for scratch space storage class must continue to work unchanged

- [x] **Understand Value and Customer Use Cases**
  - *Describe the feature's value to customers:*
    - Ensures consistent storage provisioning behavior, reducing confusion and configuration errors
    - Aligns temporary resource allocation with production storage requirements
  - *List the customer use cases identified:*

>     - As a VM owner, I want scratch space to use the same storage class as my disk so that provisioning is consistent

Reviewer: ```suggestion
    - As a VM owner, I want scratch space to use the same storage class as my volume so that provisioning is consistent
```

Author: Done.

    - **Documentation:** Documentation update is required to explain new default behavior
  - *Note any NFRs not covered and why:*
    - **UI:** Not applicable. backend storage class selection logic only, no UI changes.
    - **Monitoring:** Not applicable. Existing metrics reflect scratch space creation, no new metrics required
    - **Observability:** Not applicable. Scratch space allocation is logged through existing CDI events
    - **Performance:** Not applicable. No impact on provisioning performance
    - **Security:** Not applicable. Uses existing CDI RBAC for storage operations, no new security requirements.
    - **Scalability:** Not applicable. Selection logic scales with existing CDI controller capabilities

#### **2. Known Limitations**

None — reviewed and confirmed with Kate Shvaika, Danny Sanatar / 2026-05-05 that no feature limitations apply for this release.

#### **3. Technology and Design Review**

- [x] **Developer Handoff/QE Kickoff**
  - *Key takeaways and concerns:* Reviewed changes in Scratch Space Storage Class Selection Logic with the developer. Untestable aspects were not identified.

- [x] **Technology Challenges**
  - *List identified challenges:*

>     - Multiple disk provisioning operations can trigger scratch space allocation (clone, import, upload)

Reviewer: ```suggestion
    - Multiple volume provisioning operations can trigger scratch space allocation (import, upload)
```

Author: Done.

  - *Impact on testing approach:*

>     - Tests must cover all disk provisioning operations that use scratch space

Reviewer: ```suggestion
    - Tests must cover all volume provisioning operations that use scratch space
```

Author: Done.

- [x] **API Extensions**
  - *List new or modified APIs:*
    - No API changes - internal logic change only
    - Cluster configuration option for scratch space storage class unchanged
  - *Testing impact:*
    - No API testing updates required
    - Functional tests update required

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*
    - Ability to modify HCO configuration (for scratchSpaceStorageClass override testing)

- [x] **Topology Considerations**
  - *Describe topology requirements:* Standard cluster topology sufficient
  - *Impact on test design:* No special topology requirements

### **II. Software Test Plan (STP)**

This STP serves as the **overall roadmap for testing**, detailing the scope, approach, resources, and schedule.

#### **1. Scope of Testing**

**Testing Goals**

> - **[P0]** Verify that disk provisioning operations automatically allocate scratch space using the same storage class as the target disk when no cluster-level override is configured

Reviewer: ```suggestion
- **[P0]** Verify that volume provisioning operations automatically allocate scratch space using the same storage class as the target volume when no cluster-level override is configured
```

Author: Done.

- **[P0]** Validate that cluster administrator configuration for scratch space storage class takes priority over the default behavior

> - **[P0]** Confirm consistent scratch space storage class selection across all disk provisioning workflows (clone, import, upload)

Reviewer: ```suggestion
- **[P0]** Confirm consistent scratch space storage class selection across all volume provisioning workflows (import, upload)
```

Author: Done.

  - *Areas with reduced coverage:* None
  - *Sign-off:* Kate Shvaika/2026-05-05

**Test Environment**

- **Risk:** N/A
  - **Mitigation:** Standard test environment with multiple storage classes is sufficient for testing this feature
  - *Missing resources or infrastructure:* None
  - *Sign-off:* Kate Shvaika/2026-05-05

**Untestable Aspects**

- **Risk:** N/A
  - **Mitigation:** All scenarios can be reproduced in test environment
  - *Alternative validation approach:* None
  - *Sign-off:* Kate Shvaika/2026-05-05

**Resource Constraints**

- **Risk:** N/A
  - **Mitigation:** Current QE team capacity is sufficient for planned test execution
  - *Current capacity gaps:* None
  - *Sign-off:* Kate Shvaika/2026-05-05

**Dependencies**

- **Risk:** N/A
  - **Mitigation:** No external dependencies.
  - *Dependent teams or components:* Documentation team for behavior change documentation (non-blocking)
  - *Sign-off:* Kate Shvaika/2026-05-05

**Other**

- **Risk:** N/A
  - **Mitigation:** No additional mitigation required
  - *Sign-off:* Kate Shvaika/2026-05-05

---

### **III. Test Scenarios & Traceability**

> - **[CNV-72238]** — As a user, I want import operations to use the same storage class for scratch space as my target DataVolume

Reviewer: Not sure if you want to break this into separate scenarios but there are three different import variations that use scratch space PVCs are: 

- registry imports, 
- Http imports from unsupported server source
- Http imports of non raw files with custom certificates. 

See: https://github.com/kubevirt/containerized-data-importer/blob/main/doc/scratch-space.md

Author: Yep, I don't think I will need three different import variations.

  - *Test Scenario:* [Tier 2] Verify import requiring conversion allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want upload operations to use the same storage class for scratch space as my target DataVolume
  - *Test Scenario:* [Tier 2] Verify upload operation allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want clone operations to use the same storage class for scratch space as my source PVC
  - *Test Scenario:* [Tier 2] Verify PVC clone allocates scratch space using source PVC storage class

>   - *Priority:* P0

Reviewer: should probably remove this as clone operations don't create scratch space pvcs

Author: Ok! I was not sure about it. Thanks!
btw I've remove clone mention from other places in this STP, since it doesn't need scratch space.

# Openshift-virtualization-tests Test plan

## **Scratch Space Storage Class Selection Logic - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):**
  - [CDI Enhancement](https://github.com/kubevirt/containerized-data-importer)

>   - [HCO PR #4104](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/4104)

Author: changed

- **Feature Tracking:** https://issues.redhat.com/browse/CNV-72238
- **Epic Tracking:** https://issues.redhat.com/browse/CNV-79031
- **QE Owner(s):** Kate Shvaika (kshvaika@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage

**Document Conventions (if applicable):**
- **SC** - Storage Class
- **HCO** - HyperConverged Cluster Operator
- **CDI** - Containerized Data Importer

> - **PVC** - Persistent Volume Claim

Author: removed

### **Feature Overview**

> This feature changes the default scratch space storage class selection logic in CDI. The new behavior uses the same storage class as the target PVC for scratch space allocation instead of falling back to virt-default or default storage classes. This aligns with customer expectations that all provisioning for a particular disk should use the same storage class. The scratchSpaceStorageClass HCO configuration option continues to override this behavior when explicitly set.

Author: changed.

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

<!-- **How to complete this checklist:**
1. **Checkbox**: Mark [x] if the check is complete; if the item cannot be checked - add an explanation why in the `details` section
2. Complete the relevant, needed details for the checklist item -->

- [ ] **Review Requirements**
  <!-- Review the D/S (Downstream) requirements as defined in Jira. Understand the difference between Upstream (U/S) and D/S requirements.
  Example: "VMs must migrate without network downtime exceeding the defined threshold during node maintenance." -->
  - *List the key D/S requirements reviewed:*
    - CDI default behavior changed: scratch space now uses same storage class as target PVC
    - New logic: scratchSpaceStorageClass config -> same-as-target SC
    - Previous logic: scratchSpaceStorageClass config -> virt-default SC -> default SC
    - scratchSpaceStorageClass HCO config continues to take priority when explicitly configured

>     - Customer-visible behavior change for users not using scratchSpaceStorageClass config

Author: changed.

- [ ] **Understand Value and Customer Use Cases**
  <!-- Understand why the feature matters to customers from a D/S perspective and what the real-world use cases are.
  Example: "Customers need to migrate VMs without network downtime to maintain SLA compliance during node maintenance." -->
  - *Describe the feature's value to customers:*
    - VM owners want consistent storage class usage for all disk provisioning operations, including scratch space
    - When provisioning a disk, customers expect scratch space to use that same storage class as target
  - *List the customer use cases identified:*

>     - When provisioning a disk, scratch space should use the same storage class as that disk

Author: changed.

- [ ] **Testability**
  <!-- Confirmed requirements are **testable and unambiguous**. -->

>   - *Note any requirements that are unclear or untestable:* None - all requirements are testable through CDI operations (clone, import, upload) with various storage class configurations

Author: changed.

- [x] **Acceptance Criteria**
  <!-- Acceptance criteria are the specific, verifiable conditions that must be met for the feature to be considered complete — they define *how we know it works*.
  Example: "VM migrates without network downtime exceeding 500ms", "VM deletion is blocked for non-admin users." -->
  - *List the acceptance criteria:*
    - scratch space uses same storage class as target PVC if scratchSpaceStorageClass is not set in HCO config
    - scratchSpaceStorageClass HCO config takes priority when explicitly set
    - New selection order: scratchSpaceStorageClass config -> same-as-target SC
    - CDI functional tests verify new behavior and scratchSpaceStorageClass config override

>   - *Note any gaps or missing criteria:* None

Author: changed

Risks are *uncertainties* — things that might happen and could impact testing.
Example: "Feature does not support IPv6" is a **limitation** (known, confirmed, won't change this release).
"IPv6 cluster might not be available for testing" is a **risk** (uncertain, needs a mitigation plan). -->

<!-- Document limitations in the feature itself — constraints, trade-offs, or unsupported scenarios in the product implementation.

**Examples:**
- The feature only supports YYY storage class
- Feature does not support IPv6 (only IPv4)
- No support for ARM64 architecture in this release
- The feature is incompatible with ZZZ feature

If there are no feature limitations, remove the example items and state: "None — reviewed and confirmed with [Name/Date] that no feature limitations apply for this release." -->

- **[Feature Limitation 1]**
  - *Sign-off:* [Name/Date — confirms awareness and acceptance of this limitation]

- **[Feature Limitation 2]**
  - *Sign-off:* [Name/Date — confirms awareness and acceptance of this limitation]

Need to be updated: None identified at this time.

> 

Author: updated.

Define specific, measurable testing objectives for this feature using **SMART criteria**
(Specific, Measurable, Achievable, Relevant, Time-bound).
Each goal should tie back to requirements from Section I and be independently verifiable.

**How to Define Good Testing Goals:**
- **Specific**: Clearly state what will be tested (not "test the feature" but "validate VM live migration
  with SR-IOV networks")
- **Measurable**: Define quantifiable success criteria (e.g., "95% of VM migrations complete within xxx seconds")
- **Achievable**: Realistic given resources and timeline
- **Relevant**: Directly supports feature acceptance criteria and user stories
- **Verifiable**: Can be objectively confirmed as complete

**Priority Levels:**
- **P0**: Blocking GA - must be complete before release
- **P1**: High priority - required for full feature coverage
- **P2**: Nice-to-have - can be deferred if timeline constraints exist -->

- **[P0]** Verify scratch space uses same-as-target storage class by default (new behavior)
- **[P0]** Validate scratchSpaceStorageClass HCO config takes priority over same-as-target logic when explicitly set

> - **[P0]** Confirm all CDI operations requiring scratch space (clone, import, upload) use new selection logic

Author: changed

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.
  - *Details:* Not applicable

#### **3. Test Environment**
<!-- **Note:** "N/A" means explicitly not applicable. All items must be filled or marked N/A. -->

- **Cluster Topology:** 3-master/3-worker bare-metal
  <!-- Change if different, e.g., SNO, Compact Cluster, HCP -->

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization 4.22
  <!-- Specify exact versions to allow version traceability -->

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled
  <!-- Change only if specific CPU requirements exist -->

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM
  <!-- Adjust based on feature requirements -->

- **Special Hardware:** N/A
  <!-- Fill if needed, e.g., SR-IOV NICs, GPUs -->

- **Storage:** ocs-storagecluster-ceph-rbd-virtualization, hostpath-provisioner, custom SC
  <!-- Change if specific StorageClass(es) required -->

- **Network:** OVN-Kubernetes, IPv4
  <!-- Change if needed, e.g., Secondary Networks, IPv6, dual-stack -->

- **Required Operators:** N/A
  <!-- Add if needed, e.g., NMState Operator -->

> - **Platform:** PSIm, Bare metal

Author: changed.

#### **5. Risks**

<!-- Document specific risks for this feature. If a risk category is not applicable, mark as "N/A" with
justification in mitigation strategy. -->

**Timeline/Schedule**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Estimated impact on schedule:* N/A
  - *Sign-off:* TBD

**Test Coverage**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Areas with reduced coverage:* N/A
  - *Sign-off:* TBD

**Test Environment**

- **Risk:** N/A
  - **Mitigation:** Standard test environment is sufficient for testing this feature
  - *Missing resources or infrastructure:* N/A
  - *Sign-off:* TBD

**Untestable Aspects**

- **Risk:** Customer production storage configurations may differ from test environments
  - **Mitigation:** Focus on logic correctness rather than specific storage class types
  - *Alternative validation approach:* N/A
  - *Sign-off:* TBD

**Resource Constraints**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Current capacity gaps:* N/A
  - *Sign-off:* TBD

**Dependencies**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Dependent teams or components:* N/A
  - *Sign-off:* TBD

**Other**

- **Risk:** N/A
  - **Mitigation:** No external dependencies
  - *Sign-off:* TBD

> 

Author: updated.

  - *Test Scenario:* [Tier 2] Verify import requiring conversion allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want upload operations to use the same storage class for scratch space as my target DataVolume
  - *Test Scenario:* [Tier 2] Verify upload operation allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want clone operations to use the same storage class for scratch space as my source PVC
  - *Test Scenario:* [Tier 2] Verify PVC clone allocates scratch space using source PVC storage class
  - *Priority:* P0

- **[CNV-72238]** — As a cluster admin, I want scratchSpaceStorageClass configuration to propagate from HCO to CDI
  - *Test Scenario:* [Tier 2] Verify scratchSpaceStorageClass set in HCO CR propagates to CDI CR configuration
  - *Priority:* P0

- **[CNV-72238]** — As a cluster admin, I want to configure scratch space storage class for import operations
  - *Test Scenario:* [Tier 2] Verify import operation uses scratchSpaceStorageClass configured in HCO
  - *Priority:* P0

> 

Author: Tier 1 tests are already exist, updated by developer.
Here described just a summary of Tier 2 tests.

  Standard phases: Dev Preview (DP), Tech Preview (TP), General Availability (GA).
  For features already GA with no prior phases, only list GA.
  - GA: 4.22
- **QE Owner(s):** Kate Shvaika (kshvaika@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage

**Document Conventions (if applicable):** N/A

### **Feature Overview**

Some VM disk provisioning operations require temporary scratch space. This feature changes the default storage class selection for scratch space: it now uses the same storage class as the target disk, instead of falling back to the cluster default. Administrators can configure a specific storage class for scratch space to override this behavior.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

<!-- **How to complete this checklist:**
1. **Checkbox**: Mark [x] if the check is complete; if the item cannot be checked - add an explanation why in the `details` section

> 2. Complete the relevant, needed details for the checklist item -->

Author: fixed

- [x] **Developer Handoff/QE Kickoff**
  <!-- A meeting where Dev/Arch walked QE through the design, architecture, and implementation details. **Critical for identifying untestable aspects early.**-->
  - *Key takeaways and concerns:* Reviewed changes in Scratch Space Storage Class Selection Logic with the developer. Untestable aspects were not identified.

- [x] **Technology Challenges**
  <!-- Identified potential testing challenges related to the underlying technology.-->
  - *List identified challenges:*
    - Multiple CDI operations can trigger scratch space allocation (clone, import, upload)
  - *Impact on testing approach:*
    - Tests must cover all CDI operations that use scratch space

- [x] **API Extensions**
  <!-- Review new or modified APIs and their impact on testing. Covers both new tests for new APIs and updates to existing tests for modified APIs.
  Example: "New VirtualMachineSnapshot API v1beta2 — 3 new endpoints, 1 modified endpoint. Existing snapshot tests need updating." -->
  - *List new or modified APIs:*
    - No API changes - internal CDI logic change only
    - scratchSpaceStorageClass HCO config option unchanged

>   - *Testing impact:*

Author: changed

# Openshift-virtualization-tests Test plan

## **Scratch Space Storage Class Selection Logic - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/containerized-data-importer/pull/4054
- **Feature Tracking:** https://issues.redhat.com/browse/CNV-72238
- **Epic Tracking:** https://issues.redhat.com/browse/CNV-79031
- **Feature Maturity:**
  <!-- List each maturity phase with its target version. Use N/A for phases that don't apply.
  Standard phases: Dev Preview (DP), Tech Preview (TP), General Availability (GA).
  For features already GA with no prior phases, only list GA.

  Example:
  - DP: N/A
  - TP: 4.22
  - GA: 5.0 -->

  - DP: [version or N/A]
  - TP: [version or N/A]

>   - GA: [version]

Author: Clarifying is in progress

>   - GA: [version]

Author: Feature Maturity has been updated

- **[CNV-72238]** — As a user, I want import operations to use the same storage class for scratch space as my target DataVolume
  - *Test Scenario:* [Tier 2] Verify import requiring conversion allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want upload operations to use the same storage class for scratch space as my target DataVolume
  - *Test Scenario:* [Tier 2] Verify upload operation allocates scratch space using target DataVolume storage class
  - *Priority:* P0

- **[CNV-72238]** — As a user, I want clone operations to use the same storage class for scratch space as my source PVC
  - *Test Scenario:* [Tier 2] Verify PVC clone allocates scratch space using source PVC storage class
  - *Priority:* P0

- **[CNV-72238]** — As a cluster admin, I want scratchSpaceStorageClass configuration to propagate from HCO to CDI
  - *Test Scenario:* [Tier 2] Verify scratchSpaceStorageClass set in HCO CR propagates to CDI CR configuration
  - *Priority:* P0

- **[CNV-72238]** — As a cluster admin, I want to configure scratch space storage class for import operations

>   - *Test Scenario:* [Tier 2] Verify import operation uses scratchSpaceStorageClass configured in HCO

Author: Tier 1 tests are already exist, updated by developer.
Here described just a summary of Tier 2 tests.
